cirq-core/cirq/ops/moment.py

```diff
@@ -483,6 +483,26 @@ def _commutes_(
 
         return True
 
+    def get_single_gate_from_moment(self) -> Optional['cirq.Operation']:
+        """Returns the gate if there is only a single gate or if all gates are
+         the same, or None if the gates do not match.
+
+        Returns:
+            Operation: When there is a single operation in the Moment and
+                it is repeated one or more times, the first operation is
+                returned.
+            None: When there is no operation or different operations found, or
+                the operation is not of 'cirq.GateOperation'.
+         """
+        if not self.operations:
+            return None
+
+        for op_index in range(len(self.operations) - 1):
+            if not self.operations[op_index].equal_gates(self.operations[op_index+1]):
```

Comment:
You can the following checks to determine whether two operations have the same underlying gate. 
```python
def _have_equal_gates(op1, op2):
    if op1.gate is None or op2.gate is None:
        return False
    return op1.qubits == op2.qubits and op1.gate == op2.gate
```

This will make sure that other operations like `TaggedOperations`, `ControlledOperations`, `ParallelGateOperations` etc. work as expected when comparing the underlying gates.

Reply:
Why compare qubits?

Reply:
Oops, my bad. That's a typo. We should be comparing number of qubits, not the actual qubits. So. 
```python
def _have_equal_gates(op1, op2):
    if op1.gate is None or op2.gate is None:
        return False
    return len(op1.qubits) == len(op2.qubits) and op1.gate == op2.gate
```

This is required as of now, because of operations like `ParallelGateOperation`, which is an operation that applies an underlying single qubit gate on multiple qubits. I've also filed  #4358 to track that such an inconsistency shouldn't exist.

Reply:
Implemented.